src/main/java/net/fabricmc/fabric/mixin/client/rendercache/MixinClass853.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2016, 2017, 2018 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.client.rendercache;
+
+import net.fabricmc.fabric.api.client.model.RenderCacheView;
+import net.fabricmc.fabric.api.client.model.RenderDataProvidingBlock;
+import net.fabricmc.fabric.api.client.model.RenderDataProvidingBlockEntity;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.class_853;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.chunk.WorldChunk;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.util.Iterator;
+
+@Mixin(class_853.class)
+public abstract class MixinClass853 implements RenderCacheView {
+	private Object[] fabric_renderDataObjects;
+
+	@Shadow
+	private int field_4486;
+	@Shadow
+	private int field_4484;
+	@Shadow
+	private int field_4482;
+
+	@Inject(at = @At("RETURN"), method = "<init>")
+	public void init(World world, int cxOff, int czOff, WorldChunk[][] chunks, BlockPos posFrom, BlockPos posTo, CallbackInfo info) {
+		this.fabric_renderDataObjects = new Object[this.field_4486 * this.field_4484 * this.field_4482];
+
+		for (BlockPos pos : BlockPos.iterateBoxPositionsMutable(posFrom, posTo)) {
+			int cx = (pos.getX() >> 4) - cxOff;
+			int cz = (pos.getZ() >> 4) - czOff;
+			WorldChunk chunk = chunks[cx][cz];
+			BlockState state = chunk.getBlockState(pos);
+			fabric_renderDataObjects[method_3691(pos)] = ((RenderDataProvidingBlock) state.getBlock()).getRenderDataObject(state, world, pos);
+		}
+
+		for (WorldChunk[] chunkA : chunks) {
+			for (WorldChunk chunkB : chunkA) {
+				for (BlockEntity entity : chunkB.getBlockEntityMap().values()) {
```

Comment:
You should check the chunk coordinates here to see if they are within the region we are caching and continue if not.  Then you won't need the check below.

Cache regions are always going to be on chunk boundaries, so probably best to compute the chunk X,Z min, max before the loop starts so that you have them handy.  (Should just be the blockMin/Max >> 4)

Reply:
The chunk array passed to class_853 is already cropped to only contain the region we are caching, no?

Reply:
It would make sense for it be.  If so, then you don't need the bounds checks on the entity block position.  May still be worth doing a bounds check on the result of method_3691, but really anything out of bounds would seem to be a hard fault.

Was also thinking about your earlier comment (below.) Perhaps I misunderstood.

> And I'd still like to see benchmarks (and/or Player's opinion) to see if it's really warranted to optimize cache building like so, particularly as I believe we query a grand total of 48x256x48's worth of tile entities for an 18x18x18 right now?

Reply:
Bah, I see it now.  Render sections vs chunks.

Reply:
So, I see that you do still need to check the block entity position, but the call to getRenderDataObject() after the bounds check passes.  We don't know how expensive getRenderDataObject() will be but entity.getPos() is cheap, as are the bounds checks.